app/src/main/java/pm/gnosis/heimdall/data/repositories/impls/IpfsTransactionDetailRepository.kt

```diff
@@ -0,0 +1,134 @@
+import pm.gnosis.heimdall.data.repositories.*
+import pm.gnosis.model.Solidity
+import pm.gnosis.utils.asEthereumAddressString
+import pm.gnosis.utils.hexToByteArray
+import pm.gnosis.utils.isSolidityMethod
+import pm.gnosis.utils.removeSolidityMethodPrefix
+import timber.log.Timber
+import java.math.BigInteger
+import java.util.concurrent.ConcurrentHashMap
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Singleton
+class IpfsTransactionDetailRepository @Inject constructor(
+        authenticatorDb: GnosisAuthenticatorDb,
+        private val ethereumJsonRpcRepository: EthereumJsonRpcRepository,
+        private val ipfsApi: IpfsApi
+) : TransactionDetailRepository {
+
+    private val descriptionsDao = authenticatorDb.descriptionsDao()
+
+    private val detailRequests = ConcurrentHashMap<String, Observable<GnosisSafeTransactionDescription>>()
+
+    override fun loadTransactionDetails(descriptionHash: String, address: BigInteger, transactionHash: String?): Observable<TransactionDetails> {
+        return loadDescription(descriptionHash, address, transactionHash)
+                .map { decodeTransactionResult(descriptionHash, it) }
+                .onErrorReturnItem(TransactionDetails.unknown(descriptionHash))
+    }
+
+    private fun loadDescription(hash: String, address: BigInteger, transactionHash: String?): Observable<GnosisSafeTransactionDescription> {
+        return descriptionsDao.loadDescription(hash)
+                .subscribeOn(Schedulers.io())
+                .map { it.fromDb() }
+                .toObservable()
+                .onErrorResumeNext { _: Throwable ->
+                    loadDescriptionFromIpfs(hash, address, transactionHash)
+                            .map {
+                                descriptionsDao.insertDescription(it.toDb(hash))
+                                it
+                            }
+                }
+    }
+
+    private fun loadDescriptionFromIpfs(hash: String, address: BigInteger, transactionHash: String?): Observable<GnosisSafeTransactionDescription> {
+        return detailRequests.getOrPut(hash, {
+            createIpfsIdentifier(hash)
+                    .flatMap { ipfsApi.transactionDescription(it) }
+                    .flatMap { verifyDescription(address, transactionHash, it) }
+                    .doOnTerminate { detailRequests.remove(hash) }
+                    .publish()
+                    .autoConnect()
+        })
+    }
+
+    private fun verifyDescription(address: BigInteger, transactionHash: String?, description: GnosisSafeTransactionDescription): Observable<GnosisSafeTransactionDescription> {
+        return Observable.fromCallable<String> {
+            if (address != description.safeAddress ||
+                    transactionHash != null && description.transactionHash.removePrefix("0x") != transactionHash.removePrefix("0x")) {
+                Timber.w(IllegalStateException("Invalid description data!"))
```

Comment:
Should we enforce this and throw an exception?

If we enforce it we should exclude the "address" part, as it should be possible to get the description for other safes too